[자동차 경주 게임] 복수한다!!! 미션 제출합니다.

snowman95 opened the pull request:
안녕하세요 
복수방 복수한다!!! 입니다.

2번째 미션 "자동차 경주 게임" 제출합니다.
바쁘신 와중에 소중한 시간 내주셔서 너무나 감사드립니다.

snowman95 committed:
docs(README.md): 프로그램 요구사항 명세 작성

docs/README.md 파일
- 프로그램 요구 사항 명세 작성

snowman95 committed:
feat(RacingGame, Car): 사용자가 입력한 이름대로 자동차 객체 생성

RacingGame 의 새 기능:
- 쉼표(,) 구분으로 자동차 이름 입력받기
- 입력 받은 자동차 이름대로 Car 객체 생성

Car의 새 기능:
- 자동차 이름, 전진 횟수 정보를 가짐

snowman95 committed:
feat(play, move): 사용자가 입력한 이동 횟수만큼 레이스 게임 수행

play 의 새 기능:
- 사용자가 입력한 이동 횟수를 받아서 레이스 게임 시작

move 의 새 기능:
- 이동 횟수 만큼 자동차를 이동한 결과를 result 변수에 기록
- 0~9 사이의 랜덤 숫자를 받아서 4 이상일 경우에만 이동

snowman95 committed:
feat(showTotalRecords, showWinner): 레이싱 기록 및 우승자 출력

showTotalRecords 의 새 기능:
- n회 동안의 전체 자동차 진행 사항 출력

showWinner 의 새 기능:
- 경기가 끝났을 때 n번 전진한 자동차를 콤마로 구분하여 출력

snowman95 committed:
feat(checkNameValidation, checkRaceCountValidation): 입력 값 정합성 검사

checkNameValidation 의 새 기능:
- 자동차 이름을 배열 형태로 받아서 정합성 검사

checkRaceCountValidation 의 새 기능:
- HTML Element 에서 숫자 입력만 가능하도록 설정되어있지만, 입력이 아예 없는 경우를 검사

snowman95 committed:
feat: hidden class 추가/삭제

HTML 요소들에 hidden class 추가하여
필요한 순간에 요소가 나타나도록 제어함

snowman95 committed:
fix: 우승자 출력 element 수정

Before :
하나의 span 요소 안에 최종 우승자: 우승자1, 우승자2, 우승자3 내용 전체 출력

After :
span 요소를 2개로 나눠서 아래와 같이 따로 출력하도록 변경
1)최종 우승자: 2)우승자1, 우승자2, 우승자3

snowman95 committed:
style: 주석 추가 및 일부 스타일 변경

move 함수 주석 추가
hidden class 추가 부분 hideElement 함수로 만들어서 따로 빼냄

bigsaigon333 commented on index.html:
```diff
@@ -14,16 +14,17 @@ <h1>🏎️ 자동차 경주 게임</h1>
-      <h4>📄 실행 결과</h4>
+      <h4 id="result">📄 실행 결과</h4>
     </div>
+    <script src="https://cdn.jsdelivr.net/npm/@woowacourse/mission-utils@1.0.1/dist/mission-utils.min.js"></script>
     <script type="module" src="src/index.js"></script>
```
script tag를 body 하단에 위치하는 이유는 뭘까요?
script type이 module 인 것과 아닌 것의 차이는 무엇일까요?
script type이 module 인 경우에도 body 하단에 위치해야할까요?
한번 고민해보시면 좋을거 같아요 :smile:

snowman95 replied:
아니요. 굳이 body 하단에 위치할 필요는 없습니다.

#### script tag를 body 하단에 위치하는 이유
1. 브라우저는 [HTML 읽기→HTML 파싱→DOM트리 생성] 순서로 동작하는데 
HTML를 읽다가 script tag를 만나면 HTML 파싱을 중단하고 javascript 파일을 로드 후 javascript 코드를 파싱하므로
그 만큼 Display에 Rendering 되는 시간이 지연 될 수 있습니다.
2. DOM 트리가 생성되기전에 자바스크립트가 생성되지도 않은 DOM의 조작해버리는 문제가 생길 수 있습니다.

#### 차이점
1. 브라우저는 외부 모듈 스크립트와 기타 리소스를 병렬적으로 불러오므로 브라우저의 HTML 처리가 멈추지 않습니다.
2. 모듈 스크립트는 항상 지연 실행되므로 HTML 보다 일찍 로드되어도 HTML 문서가 완전히 준비될 때까지 대기 상태에 있다가 HTML 문서가 완전히 만들어진 이후에 실행됩니다.  
따라서 항상 완전한 HTML 페이지를 볼 수 있고 문서 내 요소에도 접근할 수 있습니다.

---
왜 이렇게 사용하는지 한번도 생각해본 적 없었는데 아주 큰 공부가 되었습니다. 😊

bigsaigon333 replied:
정리 너무 깔끔하게 잘 하셨네요~ :+1: :+1: :+1: 

추가적으로 DOMContentLoaded event, load event, module이 실행되는 시점도 한번 찾아보시면 좋아요 :+1:

bigsaigon333 commented on src/car.js:
```diff
@@ -0,0 +1,17 @@
+    this.result = [];
+  }
+
+  move = function (min = 0, max = 9, base = 4, actionCount) {
```
```suggestion
  move(min = 0, max = 9, base = 4, actionCount) {
```

작성해주신 코드와 위의 코드의 차이는 무엇일까요?
프로토타입 체인 상에서 어떻게 다른지 알고 계실까요..? :smile:

snowman95 replied:
### 문법 상으로는 아래와 같은 차이가 있습니다.

#### 함수 표현식으로 정의된 인스턴스 프로퍼티(변수)
```diff
- move = function (min = 0, max = 9, base = 4, actionCount) {
``` 
#### 함수 선언식으로 정의된 클래스 프로토타입 메소드(함수)
```diff
+ move(min = 0, max = 9, base = 4, actionCount) {
``` 
---
### 클래스의 개념 상으로는 아래와 같은 차이가 있습니다.
설령 자바스크립트가 함수를 변수에 담을 수 있고 그것이 기능적으로 잘 동작하더라도,
move(차를 움직여라) 라는 기능은 차의 속성,개성이 아닌 행동(Action)이므로 메소드(함수)로 선언하는 것이 올바른 방법인 것 같습니다. 

---
### 프로토 체인 상에서의 차이는 아래와 같습니다.
#### 아래 코드는 인스턴스 프로퍼티입니다. 클래스 생성자 내부에 선언한 것과 동일합니다. 

생성자 초기화될 때 정의되며 prototype 내부에 없습니다.
prototype 내부에 없기 때문에 프로토타입 체인이 일어나지 않습니다.

```diff
- move = function (min = 0, max = 9, base = 4, actionCount) {
``` 
```
Car {name, moveCount, result}
  move : f move(min, max, base, actionCount) {
    prototype : {
      constructor : f Object()
     }
  }
  name : 
  moveCount : 
  result : 
  prototype = {
    constructor : class Car
  }
```

#### 아래 코드는 클래스 프로토타입 메소드 입니다.
Car 클래스의 prototype 객체 안에 들어갑니다.
프로토타입 체인을 이용한 상속을 제대로 활용하기 위해서는 이와 같은 프로토타입 메소드로 기능을 정의해야 합니다.

```diff
+ move(min = 0, max = 9, base = 4, actionCount) {
``` 
```
Car {name, moveCount, result}
  name : 
  moveCount : 
  result : 
  prototype = {
    constructor : class Car
    move: f move(min, max, base, actionCount)
  }
```
 
---

피드백 해주신 것 보고 내가 왜 이렇게 썼지 하고 아차 싶었습니다...
오히려 그 덕분에 정말 중요한 내용을 공부할 수 있었네요..
제가 공부하고 이해한 내용을 적어보았습니다. 조금이라도 잘못된 부분은 회초리 부탁 드립니다.

bigsaigon333 replied:
> 제가 공부하고 이해한 내용을 적어보았습니다. 조금이라도 잘못된 부분은 회초리 부탁 드립니다.

굿입니다~  :+1: :+1: :+1: :+1:  

단순히 객체의 프로퍼티가 함수인 것과 메서드의 차이를 명확히 이해하신 것 같아서 좋습니다~

bigsaigon333 commented on src/tools.js:
```diff
@@ -0,0 +1,31 @@
+  let valid = false;
+  let message = "";
+  const size = inputArr.length;
+  if (inputArr.filter((item) => item.length === 0).length > 0) {
```
`Array.prototype.some` 또는 `Array.prototype.every` 메서드를 써보는건 어떨까요?

snowman95 replied:
Array.prototype.some() : 배열의 어느 한 요소라도 조건에 부합하면 true 반환
Array.prototype.every() : 배열의 모든 요소가 조건에 부합하면 true 반환

---

제가 딱 필요로 하던 기능입니다.
자바스크립트는 정말 기능이 너무 많네요 🤣
```diff
- if (inputArr.filter((item) => item.length === 0).length > 0) {
+ if (inputArr.some((term) => term.length === 0)) {
```

bigsaigon333 replied:
mdn 에서 string이나 Array는 관련 메서드를 심심할때마다 봐두고, 
이거 쓰면 될거 같은데? 같은게 있으면 바로 써보면 안까먹게 되더라구요~

bigsaigon333 commented on src/tools.js:
```diff
@@ -0,0 +1,31 @@
+    valid = true;
+  }
+
+  return { valid, message };
```
`{valid, message}` 가 아니라 반환값이 `""` 일 때는 valid 를 의미하고, 반환값이 `""` 이 아닐 때는 invalid 를 의미하게끔 하는 것도 많이 쓰이는 방식입니다 :smile: 

https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/validationMessage

snowman95 replied:
네 valid 값을 굳이 쓸 필요 없더군요 ㅎㅎ
기능이 유효성 검사라서 true/false로 반환해야 한다는 생각에 사로잡혀 있었던 것 같습니다. 😮

bigsaigon333 commented on src/tools.js:
```diff
@@ -0,0 +1,31 @@
+  let message = "";
+  if (input.length === 0) {
+    message = "error code1: 입력값이 없습니다.";
+  } else if (typeof Number(input) === NaN) {
```
`typeof NaN` 은 `number` 이므로, 해당 조건은 절대 충족되지 않겠네요 :smile: 

어떻게 하면 `NaN` 을 특정할 수 있을까요?

snowman95 replied:
사용자가 입력 창에 숫자를 넣어도 문자열 형태로 가져오기 때문에
Number로 변환하여 숫자인지 확인하려고 넣었는데 지금 보니 말도 안되는 구문이었네요...

HTML 단에서 input type="number"를 통해 숫자만 입력되도록 제한해 두어서 위 조건문은 제외하는 것으로 수정했습니다.

만약 input type="text" 였다면
문자열의 모든 요소가 "0123456789" 중 하나 인지 검사하는 방법을 쓸 수 있었을 것 같습니다.

bigsaigon333 replied:
> 만약 input type="text" 였다면
문자열의 모든 요소가 "0123456789" 중 하나 인지 검사하는 방법을 쓸 수 있었을 것 같습니다.

`1e6` 과 같은 지수표기법이나 `-3`과 같은 음수도 JS상 유효한 숫자 표기법이기 때문에, 
위와 같은 방법으로 손쉽게 걸러낼 수는 없을거에요 :smile: 

심오하고도 복잡한 input과 form의 세상이지만, 말씀하신 방법만으로도 충분하다고 생각합니다 :+1: 

참고로 제가 그냥 말씀드리고 싶었던건 `Number.isNaN` 이라는 메서드를 통해서 `NaN` 을 특정가능하다는 거 말하고 싶었어요
`NaN === NaN` 은  신기하게도 `false`  이거든요 :laughing:

snowman95 replied:
오 새로운 것 알아갑니다 ㅠㅠ 아주 간단하면서도 강력한 함수가 있군요!!
`Number.isNaN` 꼭 기억해두었다가 써먹어야겠습니다.

bigsaigon333 commented on src/tools.js:
```diff
@@ -0,0 +1,31 @@
+    message = "error code1: 입력값이 없습니다.";
+  } else if (inputArr.filter((item) => item.length > 5).length > 0) {
+    message = `error code2: 자동차 이름은 5자 이하만 가능합니다.`;
+  } else if ([...new Set(inputArr)].length < size) {
```
`size`를 굳이 위에 상수로 정의할 필요가 있었을까 싶네요 :smile: 

9행에서 한번만 사용되고 있기도 하고, 
`size`의 정의는  4행에 있는데 실제 사용은 9행이라서 간격이 좀 있어서 아래 위로 왔다 갔다 해야하는 것도 번거롭고, 
`size` 라는 이름이 너무 추상적이기도 한 거 같아요

만약 4행과 9행 사이에서 `inputArr` 에 요소를 추가한다면, (그러면 안되겠지만 :angry: )
실제 `inputArr.length` 와 `size`는 다른 값이 되어서 버그가 발생할 가능성도 있을 거 같아요 :smile:

snowman95 replied:
맞습니다. 불필요한 상수였습니다.
사실 C/C++로 코딩하던 옛 습관이 따라온 것 같긴한데 변화에 적응해보겠습니다..

bigsaigon333 replied:
C++ 해봤던게 너무 옛날이라 잘 기억은 안나지만, 
JS에서는 배열의 length가 단순히 property 이기 때문에 O(1) 으로 배열 크기를 알 수 있어서 
`배열.length` 이렇게 많이 쓰는거 같아요 :+1: 

message 도 그냥 return 시켜버렸어도 좋을것 같네요~

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,114 @@
+
+  showWinner = function () {
+    const winner = [];
+    this.cars.forEach((car) =>
+      car.result[this.raceCount - 1] === this.raceCount
+        ? winner.push(car.name)
+        : ""
+    );
```
forEach 의 보조함수가 너무 많은 역할을 하고 있는 것 같아요. 
그냥 for문을 쓰는 것과 비교해 크게 다를 바가 없는 것 같아요. 
3항 연산자를 이용해서 표현식을 반환하는 것은 forEach문의 보조함수의 제대로 된 사용법은 아닌 것 같아요.
아무 것도 수행하지 않기 위해 `""`을 반환하기 보다는 `&&` 연산자를 사용하는게 오히려 나았을 수도 있지만, 
표현식을 쓰는거 자체가 여기서 좋은 방법은 아닌거 같습니다 :smile: 

그냥 for-of 문을 쓰거나, 
함수형으로 작성하고 싶다면 보조 함수가 하나의 일만 수행하게 단순하게 적는 것이 좋습니다 :smile: 

```js
for(const car of this.cars) {
  if (car.result[this.raceCount - 1] === this.raceCount) {
    winner.push(car.name);
  }
}
```

```js
const winner = this.cars
  .filter((car) => car.result[this.raceCount - 1] === this.raceCount)
  .map(({name}) => name)
```

snowman95 replied:
네 filter 함수로 훨씬 간결하고 깔끔하게 구현할 수 있었군요!
덤으로 선생님의 고급기술 ({name}) 도 배워갑니다...

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,114 @@
+        : ""
+    );
+    this.resultText.innerHTML = `${winner.join(", ")}`;
+    this.winnerOuputElement.appendChild(this.resultTitle);
+    this.winnerOuputElement.appendChild(this.resultText);
```
```suggestion
    this.winnerOuputElement.append(this.resultTitle, this.resultText);
```

`appendChild` 와 `append` 의 차이는 무엇일까요?

snowman95 replied:
append : 노드 객체, 문자열 모두 추가할 수 있고 여러 개 한번에 추가 가능
appendChild : 노드 객체만 추가할 수 있고 하나만 추가 가능

---

같은 부모에 추가하는 것이니 append로 한꺼번에 추가하는 것이 더 좋았겠군요! ㅎㅎ

bigsaigon333 replied:
>  문자열 모두 추가 
:+1: :+1: :+1: :+1: 

추가로 `appendChild` 에서는 문자열을 어떻게 추가하지? 라는 생각도 한번 해보셨으면 좋겠네요 :+1:

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,114 @@
+    this.winnerOuputElement.classList.remove(HIDDEN_CLASSNAME);
+  };
+
+  getRaceRecord = function (lap) {
+    const record = [];
+    this.cars.forEach((car) =>
+      record.push(`${car.name}:${" -".repeat(car.result[lap])}`)
+    );
+    record.push("<br/>");
+    return record.join("<br/>");
+  };
```
[아래의 showWinner 리뷰를 참고해주세요 :smiley: ](https://github.com/REVENGE-WOOTECO/javascript-racingcar-precourse/pull/1/files/2db462097688832b5c3fcc378c2d4c5e24ae6259#r762501444)

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,114 @@
+    event.preventDefault();
+    const raceCount = this.racingCountInput.value;
+    const { valid, message } = checkRaceCountValidation(raceCount);
+    valid ? this.play(raceCount) : alert(message);
```
표현식으로 평가된 값을 사용하지 않는 것은  좋은 코드가 아닙니다. 
다른 개발자가 볼 때 평가된 값이 사용되지 않는 것이 로직 에러로 보이기 때문이죠
`this.play` 와 `alert` 모두 값을 반환하지 않고 사이드 이펙트만 생성하고 있습니다.
그냥 if문을 쓰시는게 더 의도를 명확하게 보여줄 수 있습니다.

https://eslint.org/docs/rules/no-unused-expressions

snowman95 replied:
이 부분은 제가 고민을 꽤 많이 했는데
결국 아래와 같이 수정해 보았습니다.

유효성 검사를 통해 error가 발견되면 error code를 넘겨받고
error 가 존재할 시 error code에 해당하는 문자열을 받아서 alert() 후 프로그램 흐름을 return 으로 끊기
- checkRaceCountValidation() : error 존재시 code를 반환. 없으면 ""
- getMessageByErrorCode() : error code에 해당하는 문자열 반환

```
const error = checkRaceCountValidation(raceCount);
    if (error) {
      alert(getMessageByErrorCode(error));
      return; → 프로그램 흐름 중단
    }
this.play(raceCount); → error 없으면 그대로 실행
```
![image](https://user-images.githubusercontent.com/89284542/145198683-c4be1d93-ca4f-4aff-bd80-f9f98d304d78.png)
![image](https://user-images.githubusercontent.com/89284542/145198716-3fde110a-0adf-48d0-a843-9cdd602c9136.png)

---

제가 indent depth 가 카운트되는 방식을 잘못 이해하여 
줄곧 클래스(1) - 메소드(2) - 끝 이렇게 알고 있었네요 ㅠㅠ😣😣😣
제가 심하다 싶을 정도로 삼항 연산자와 forEach 등에 집착하던 이유가 이 때문이었습니다..

bigsaigon333 replied:
> 제가 indent depth 가 카운트되는 방식을 잘못 이해하여
줄곧 클래스(1) - 메소드(2) - 끝 이렇게 알고 있었네요 ㅠㅠpersevereperseverepersevere
제가 심하다 싶을 정도로 삼항 연산자와 forEach 등에 집착하던 이유가 이 때문이었습니다..

아 그런 제약조건 때문에 그러셨군요.. 생각도 못했네요
사실 저희 때도 한 메서드 15줄, depth 2  이런 제약이 있었던 거 같네요 :dizzy_face: 

근데 제약조건을 맞추기 위해 메서드(또는 함수)를 최대한 분리하라는게 아마 우테코의 의도아닐까 싶어요 :smile:

bigsaigon333 replied:
```js
const error = checkRaceCountValidation(raceCount);
    if (error) {
      alert(getMessageByErrorCode(error));
      return; → 프로그램 흐름 중단
    }
this.play(raceCount); → error 없으면 그대로 실행
```

:+1: :+1: :+1: :+1:

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,114 @@
+    event.preventDefault();
+    const carNames = this.carNameInput.value.split(",");
+    const { valid, message } = checkNameValidation(carNames);
+    valid ? this.makeCarWithName(carNames) : alert(message);
```
[65번째 리뷰를 참고해주세요 :smile: ](https://github.com/REVENGE-WOOTECO/javascript-racingcar-precourse/pull/1/files/2db462097688832b5c3fcc378c2d4c5e24ae6259#r762502130)

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,114 @@
+import { RACE_GAME_INPUT } from "./constants/race.js";
+import { HIDDEN_CLASSNAME } from "./constants/style.js";
+import { checkNameValidation, checkRaceCountValidation } from "./tools.js";
+export default class RacingGame {
```
왜 class 를 쓰셨을까요?
싱글턴으로 사용하고 있어서 재사용성을 고려해서 사용 한 건 아닌것 같네요. 
또한 이 객체가 굉장히 많은 프로퍼티를 가지게 되어서 메서드에서 모든 프로퍼티에 다 접근할 수 있겠네요 :smiley_cat: 

- 꼭 class로 쓰는게 맞을까? 
- class의 원래 목적은 무엇일까?
- 내가 추가하게 되는 메서드에서 모든 프로퍼티에 접근할 수 있는게 좋을걸까? 

이런 걸 한 번쯤 고민해보셨으면 좋겠습니다. 

저도 작년에 이런식으로 코드를 작성하였지만, 
되돌아봤을 때 그냥 다른 언어에도 있던 class 가 먼가 익숙해서 아무 이유없이 썼던 것 같아서, 
그러지 않으시길 바래 말씀드립니다 :smile:

snowman95 replied:
이것이 과제/미션이 아니라 실제 프로젝트였다면 아래와 같이 구현하는 것이 정답이라고 생각합니다.

- Car : 이름만 다르고 동일 기능하는 인스턴스를 찍어내야 하므로 함수가 아닌 class로 구현
- RacingGame : 게임의 전체 흐름을 관리하는 게임 관리자의 역할을 하는 유일한 존재이므로 재활용이 되면 안되므로 싱글톤 패턴의 class로 구현


그리고 UI 적인 요소 (HTML 객체)를 관장하는 클래스도 따로 존재하는 게 실제로는 맞을 것 같긴하네요..😅
프로그램 규모가 작다고 해서 너무 안일한 생각으로 가볍게 코드를 짠 스스로를 반성하게 됩니다... 흑흑 감사합니다.

bigsaigon333 replied:
한 번 고민해보셨다는게 중요한거 같아요~

사실 저보고 자동차 경주 게임을 vanillaJS 로 해보라고 해도, 어떻게 해야할지, 무엇이 전형적인 패턴인지 모르겠어요 :dizzy: 

화이팅화이팅!!

snowman95 replied:
네 설계적인 고민은 이미 잘 만들어진 기존 사례를 참고하는 것이 최고인 것 같아요ㅎㅎ
인기 있는 프레임워크의 구조에서도 힌트를 얻고, 디자인 패턴 사용 사례에서도 힌트를 얻고...

옛날에 지인 중 한 명은 미니 프로젝트를 할 때마다 꼭 새로운 디자인 패턴을 적용해보려고 시도하더군요.
그러곤 그 디자인 패턴을 더 빠르고 간편하게 구현할 수 있도록 좀 더 범용적인 형태의 템플릿 클래스로 변환해서 
자기만의 라이브러리에 추가했어요.
(React로 치자면 Hook 정도가 되겠네요)

그 이후의 프로젝트에서 그것들을 활용하여서 훨씬 더 빠르고 더 깔끔하게 코드를 구현하는 것을 보면서
이게 진정한 발전이고 생산성을 높이는 것이구나 하고 큰 깨달음을 느꼈습니다.
단순히 코드만 백업해두고 기억이 안 나면 부분 부분 참고하는 것과는 전혀 다른 차원이었고
프로젝트를 진행하고 코드를 적을 때마다 자기 자산이 불어나는 매직이었죠.

갑자기 생각나서 끄적끄적 해봤네요 ㅎㅎ
동동님도 참고하면 좋을 것 같아서 적어봤습니다.

bigsaigon333 replied:
좋은 말씀 감사합니다! 
이것저것 시도는 많이 해보고 있는데, 한 군데에 모을 생각은 안해봤네요!
한번 해봐야겠습니다 :+1: :+1:

bigsaigon333 reviewed (changes requested):
@snowman95 님 안녕하세요 우테코 3기 동동입니다 :smile: 

작년에 저도 자동차 경주 게임 열심히 푸는 입장이었는데, 
이제는 제가 리뷰를 하고 있으니 되게 민망하기도 하고 부끄럽기도 하네요 

일단 코드 위주로 리뷰를 봤어요. 
고민해보셨으면 좋을 포인트들을 리뷰로 남겼으니 한번 봐주시면 좋겠습니다.
학습하시면서 고민했던 내용들, 찾아봤던 내용들을 개인 노션등에 정리해두면
설령 지금 다 이해를 못하더라도, 나중에 굉장히 큰 자산이 되실거라고 생각합니다.
(제가 그러지 못했기에 드리는 말씀... :sweat_drops: )

실제 실행되는 웹페이지는 보지 못했어요. 
github pages 에 배포를 한 링크를 남겨주시면 실행된 웹페이지를 보면서 더 추가적인 리뷰 남기도록 할게요

그리고 제가 남기는 리뷰를 받아들이고 안 받아들이고는 @snowman95 마음입니다 :smile: 
제가 말한게 무조건 맞는 것도 아니니까, 한 명의 개발자의 의견이라고 생각하시고, 비판적으로 수용해주세요 :laughing:

화이팅입니다~~ :+1:

bigsaigon333 commented:
리뷰를 다 반영(?) 하셨으면 다시 리뷰 요청을 눌러주세요 :+1:

snowman95 commented:
동동 선생님...
질문 하나하나에 정말 깊은 뜻이 담겨 있군요.
하루 이틀 만에 깨우치기에 큰 어려움이 있습니다.. 조금만 기다려주세요...😅

snowman95 committed:
refactor: PR#1 피드백 사항 반영

PR #1 피드백 사항 반영 :
- 인스턴스 프로퍼티 → 클래스 프로토타입 메소드로 수정
- Array.prototype.some 함수를 활용
- {valid, message} → {message} 하나로 동작하도록 수정
- else if(typeof Number(input) === NaN) 불필요 조건문 제거
- 상수로 정의된 size 변수 제거
- forEach+삼항연산자 → filter+map 함수로 변환
- appendChild → append 함수를 이용
- valid ? this.play(raceCount) : alert(message); → 삼항연산자 대신 if문 사용
- ui 관련 부분을 별도 클래스로 떼어내서 racingGame에 import하도록 변경
- 외부에서 프로퍼티 접근하지 못하도록 private 부여

snowman95 commented:
https://snowman95.github.io/javascript-racingcar-precourse/

코드 수정 후 배포하였습니다 😊

bigsaigon333 reviewed (approved):
![image](https://user-images.githubusercontent.com/31029000/145815235-1b95b877-0e89-4cb7-8e42-5ba669dd4f95.png)

자동차 이름을 입력하는 input 의 확인 버튼을 여러번 누를 수 있네요(엔터키를 여러번 누를 수 있네요)

그랬더니 동일한 이름의 자동차가 여러대 생겨버렸어요 :dizzy_face: 

FE는 결과물이 화면으로 보여지기 때문에, 화면 단에서 테스트를 다양하게 해보시는 걸 추천드려요!

(그걸 편하게 하려고 E2E 테스트를 하는거구요)


일반적으로 form에 submit event 가 발생하면, button 을 disabled 하는게 좋습니다.
왜냐하면 저같은 이상한 유저들이 여러번 버튼을 클릭해서 submit event가 여러번 발생할 수 있기 때문이죠 :smile: 

<https://web.dev/sign-up-form-best-practices/>

일단 리뷰가 너무 길어지는 것 같아서 Approve 드릴게요~ 고생하셨습니다 다음 미션에서 뵈어요~
